Task 1.5 Implemented ProductDao

HidJaba-hub opened the pull request:
Task 1.6: Use ProductDao on PLP

(cherry picked from commit d84854e3757711f6f12df868d3a1f5065f7fac63)

HidJaba-hub committed:
Task 1.5 Implemented ProductDao
Task 1.6: Use ProductDao on PLP

(cherry picked from commit d84854e3757711f6f12df868d3a1f5065f7fac63)

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class ArrayListProductDao implements ProductDao {
```
class name should not contain any details about the functional implementation
now you are using ArrayList as a storage, what if you decide to switch to another collection? Class name should describe the purpose, here we created a DAO for handling products
Name should be smth like "CustomProductDao" or "ProductDaoImpl"

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+    private final Lock writeLock = lock.writeLock();
+    private List<Product> products;
+
+    public ArrayListProductDao() {
```
DAO should be a singleton
Otherwise if you have 10 servlets which require this DAO you will create 10 instances, which is not fine

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public Product getProduct(Long id) {
+        readLock.lock();//�� ������� ��� �� ������ ����� ���, �� ��� ������� ������������ ������ �����
```
it makes sense to create a lock for the read too

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public Product getProduct(Long id) {
```
return Optional

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Override
+    public Product save(Product product) {
```
why do you return object type? It should be void

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+            if (product.getId() != null) {
+                products.set((int) (product.getId()-1), product);
+            } else {
+                product.setId((long) products.size() + 1);
```
UUID for id generation
DAO must not store any manipulations on "Product" POJO, all initializations of Product must be defined in its own class, in constructor

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Lock readLock = lock.readLock();
+    private final Lock writeLock = lock.writeLock();
+    private List<Product> products;
```
getter, as a result you will be able to get elements for comparison in test class

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ProductDao.java:
```diff
@@ -1,10 +1,15 @@
     List<Product> findProducts();
-    void save(Product product);
+
+    Product save(Product product);
```
Optional

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ProductDao.java:
```diff
@@ -1,10 +1,15 @@
     Product getProduct(Long id);
+
     List<Product> findProducts();
-    void save(Product product);
```
revert to void

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+        readLock.lock();
+        try {
+            return products.stream()
+                    .filter(product -> product.getPrice() != null)
```
what if price is set to zero? Customer should not see such products

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/entity/Product.java:
```diff
@@ -0,0 +1,38 @@
+        this.price = price;
+        this.currency = currency;
+        this.stock = stock;
+        this.imageUrl = imageUrl;
```
UUID here

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+package com.es.phoneshop.model.product.DAO;
```
dao package outside model

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,41 +1,57 @@
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        request.setAttribute("products", getSampleProducts());
+        request.setAttribute("products", productDao.findProducts());
```
in general I suggest to add one more layer - service (for example sigleton ProductService, place it in new package service)
in ProductService we will call ProductDao, throw all possible exceptions
And as a result you can use this service in servlet and you also will be able to catch exceptions from service in servlet

Servlet calls Service -> Service calls Dao
Service is also singleton

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,12 +1,19 @@
 
-public class ArrayListProductDaoTest
-{
+import static org.junit.Assert.*;
```
no wildcards

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -16,6 +23,35 @@ public void setup() {
     @Test
     public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+        Product product = productDao.save(new Product("test", "test", new BigDecimal(100), usd, 0, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg"));
```
when you test  DAO you should not rely on it's other methods
you should mock this method save() and call only .findProducts()

if your save() is broken you will have x2 failed tests: all for save() and tests where this save() is called before the test method

esSoftPalinaTsi replied:
save the collection in setUp() to dao and just get its Product instances
no need to call save everywhere

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -16,6 +23,35 @@ public void setup() {
     public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+        Product product = productDao.save(new Product("test", "test", new BigDecimal(100), usd, 0, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg"));
+        assertTrue(product.getId() > 0);
```
tests are splitted into 3 sections
1st - "mocked resources, when().then()"
2nd - "call for the test method"
3rd - "verify(), asserts(), check the results"

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -16,6 +23,35 @@ public void setup() {
+    }
+
+    @Test
+    public void testDeleteProduct() {
```
lets follow naming convention "givenalidProduct_whenDelete_thenDeleteProduct"

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -16,6 +23,35 @@ public void setup() {
 
```
where are the tests when NoSuchElementException can be thrown? While delete, getProduct

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/product/DAO/ArrayListProductDao.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.ArrayList;
+import java.util.Currency;
+import java.util.List;
+import java.util.NoSuchElementException;
```
can be thrown also in delete

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/web/ProductListPageServletTest.java:
```diff
@@ -1,34 +1,36 @@
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
```
no wildcards as well

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -8,16 +8,30 @@
-      <tr>
+      <c:if test="${product.isChosen==true}"><tr style="background-color: aqua"></c:if>
+      <c:if test="${product.isChosen==false}"><tr></c:if>
+        <td>${product.id}</td>
```
id should not be visible to customer
you can make it a hidden field

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,41 +1,57 @@
+            case ("findProduct"):
+                if (request.getParameter("phoneId") != null && !Objects.equals(request.getParameter("phoneId"), "")) {
+                    try {
+                        productDao.getProduct(Long.valueOf(request.getParameter("phoneId"))).setIsChosen(true);
```
again - servlet must not manipulate on Product POJO
If you want to add some moer functionlity, then you need to call service in servlet
service calls dao
dao sets this attribute

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,41 +1,57 @@
+                }
+                break;
+            case ("deleteProducts"):
+                if (request.getParameter("phoneIdToDelete") != null && !Objects.equals(request.getParameter("phoneIdToDelete"), "")) {
```
everywhere
no need to call request.getParameter("phoneIdToDelete") twice
please move it to variable

esSoftPalinaTsi replied:
everywhere
StringUtls.isEmpty()

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,41 +1,57 @@
+                break;
+            case ("deleteProducts"):
+                if (request.getParameter("phoneIdToDelete") != null && !Objects.equals(request.getParameter("phoneIdToDelete"), "")) {
+                    productDao.delete(Long.valueOf(request.getParameter("phoneIdToDelete")));
```
customer must not delete products 
only buy :)

HidJaba-hub replied:
May i change naming of this option into "Buy product"? But it will still work as deleting

HidJaba-hub committed:
-added service
-service/dao are singleton now
-fixed tests
-added string search for products

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/DAO/CustomProductDao.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public static CustomProductDao getCustomProductDao() {
+        CustomProductDao localInstance = customProductDao;
```
no need for additional localInstance , just check and set customProductDao

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/DAO/CustomProductDao.java:
```diff
@@ -0,0 +1,137 @@
+        writeLock.lock();
+        try {
+            if (product == null)
+                throw new NullPointerException();
```
null pointer is too abstract, here it makes sense to add a custom exception

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/StringChecker.java:
```diff
@@ -0,0 +1,14 @@
+
+public class StringChecker {
+    public static boolean containsWords(String string, String search) {
+        if (string == null || search == null) {
```
StringsUtils instead of nulls

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/StringChecker.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.Arrays;
+
+public class StringChecker {
```
not critical, but FYI: when class contains only boolean check it makes sense to implement Predicate<T>

esSoftPalinaTsi replied:
why do we need this in the first task?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+    private static final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private static final Lock writeLock = lock.writeLock();
+    //idk is it necessary to synchronize service, because there are read/write locks in dao and it will be locked for any operation
+    private static volatile CustomProductService customProductService;
```
why volatile

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+public class CustomProductService implements ProductService {
+    private static final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private static final Lock writeLock = lock.writeLock();
+    //idk is it necessary to synchronize service, because there are read/write locks in dao and it will be locked for any operation
```
no need, you are right here
syncs are needed when the data can be modified
data modification is the aim of dao

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+    private final ProductDao productDao;
+
+    private CustomProductService() {
+        productDao = CustomProductDao.getCustomProductDao();
```
getInstance() for getting the singleton

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public static CustomProductService getCustomProductService() {
+        CustomProductService localInstance = customProductService;
```
same as for dao with localInstance

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+            writeLock.lock();
+            try {
+                localInstance = customProductService;
+                if (localInstance == null) {
```
no need in if inside if

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<Product> getProductsNotNull();
+
+    void buyProduct(Long id);
```
what does it mean in scope of 1st task
and why buying of the product means its deletion? Deletion of product mean that you remove it from  database(in your case from list) and this product is not present for others customers
Buying means decreasing of the stock

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+public interface ProductService {
+    Product getProductById(Long id);
+
+    List<Product> getProductByDescription(String description);
```
what does it mean in scope of 1st task

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<Product> getProductByDescription(String description);
+
+    List<Product> getProductsNotNull();
```
NotNull should not be present in method name

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+
+    void buyProduct(Long id);
+
+    void changeState(Product product, boolean state);
```
what does it mean in scope of 1st task

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    public Product getProductById(Long id) {
+        Product product = null;
```
we should never return  nulls, it's dangerous

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+        try {
+            Optional<Product> optionalProduct = productDao.getProductById(id);
+            if (optionalProduct.isPresent()) {
+                product = optionalProduct.get();
```
just return optional.get(), without additional variable

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,81 @@
+    public List<Product> getProductByDescription(String description) {
+        List<Product> products;
+
+        products = productDao.getProductByDescription(description);
```
no need in local variable, just return is enough

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+
+public interface ProductService {
+    Product getProductById(Long id);
+
```
save() is missed

esSoftPalinaTsi commented on src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -10,19 +10,18 @@
 
-      <p>Найти товар по номеру: <input type="text" placeholder="Номер позиции" name="phoneId"></p>
-      <button type="submit" value="findProduct" name="action">Найти товар</button>
+      <p>Найти товары по названию: <input type="text" placeholder="Название товара" name="phoneDescription"></p>
```
this is from the 2 task, isn't it?

esSoftPalinaTsi replied:
each action is a separate request (get/post)

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
+public class ArrayListProductDaoTest {// i suppose service tests must be also written
     Currency usd = Currency.getInstance("USD");
-    private ProductDao productDao;
+    ProductDao productDao;
```
private modifiers are missed from all variables

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
 
-public class ArrayListProductDaoTest {
+@RunWith(MockitoJUnitRunner.class)
+public class ArrayListProductDaoTest {// i suppose service tests must be also written
```
exactly, all the code is required to be covered

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
+
+    @After
+    public void rearrange() {
+        productList.remove(product);
```
what is happening here?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/model/entity/Product.java:
```diff
@@ -28,6 +31,7 @@ public class Product {
     private Boolean isChosen = false;
 
     public Product(String code, String description, BigDecimal price, Currency currency, int stock, String imageUrl) {
+        this.id = (long) UUID.randomUUID().hashCode();
```
no sense in convertation of int to long for id
 getMostSignificantBits()

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
         assertFalse(products.isEmpty());
         assertFalse(products.stream()
-                .anyMatch(product1 -> product1.equals(product)));
+                .anyMatch(product_ -> product_.equals(product)));
```
product_?

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
-    public void testFindProductsNoResults() {
-        Product product = productDao.save(new Product("test", "test", new BigDecimal(100), usd, 0, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg"));
-        assertTrue(product.getId() > 0);
+    public void givenAProductWithZeroStock_whenFindProductsWithoutZeroStock_thenCheckExistenceOfGivenProduct() {
```
can be cut up to givenProductWithZeroStock_whenFindProducts_thenGetProducts
all other parts are obvious

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
         List<Product> products = productDao.findProducts();
+
         assertFalse(products.isEmpty());
         assertFalse(products.stream()
```
stream() can be replaced with simple list.contains()

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
+
+    @Test
+    public void givenAProduct_whenGetProductById_thenGetThisProduct() {
+        Product gottenProduct = productDao.getProductById(product.getId()).get();
```
expectedProduct

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:
```diff
@@ -1,57 +1,118 @@
+        Product gottenProduct = productDao.getProductById(product.getId()).get();
+
+        assertNotNull(gottenProduct);
+        assertSame(product, gottenProduct);
```
what is the difference between assertSame and assertEquals

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,57 +1,55 @@
-                        exception.printStackTrace();
-                    }
+                phoneDescription = request.getParameter("phoneDescription");
+                if (phoneDescription != null && !phoneDescription.isEmpty()) {
```
the replacement to StringUtils comment was not fixed

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,57 +1,55 @@
     }
 
     @Override
     protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
is it from the 2 lecture?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:
```diff
@@ -1,57 +1,55 @@
-        productDao.findProducts().forEach(product -> product.setIsChosen(false));
+        productService.getProductsNotNull().forEach(product -> productService.changeState(product, false));
         switch (action) {
             case ("findProduct"):
```
findProduct is not doPost

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/ProductService.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.util.List;
+
+public interface ProductService {
```
+ unit test

HidJaba-hub committed:
Fixed some features:
-added custom runtime exception class
-added service tests
-stringChecker is biPredicate

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/exception/ProductException.java:
```diff
@@ -0,0 +1,11 @@
+package com.es.phoneshop.exception;
+
+public class ProductException extends RuntimeException {
```
name can be more detailed, not clear which exactly product issue is expected to happend

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -14,68 +15,62 @@
+        if (optionalProduct.isPresent()) {
+            return optionalProduct.get();
+        } else {
+            throw new NoSuchElementException("Product not found for id: " + id);
```
is your custom ProductException done for this?

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static org.junit.Assert.*;
```
wilcard

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+import static org.mockito.Mockito.doAnswer;
+
+@RunWith(MockitoJUnitRunner.class)
+public class ArrayListProductDaoTest {
```
test class name

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+        List<Product> products = productDao.findProducts();
+
+        assertFalse(products.isEmpty());
+        assertFalse(products.contains(product));
```
by your code I see that this is impossible, you're filtering zero stock

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+        List<Product> products = productDao.findProducts();
+
+        assertFalse(products.isEmpty());
+        assertFalse(products.contains(product));
```
by your code I see that this is impossible, you're filtering nullable price

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+        List<Product> productsAfterDel = CustomProductDao.getInstance().getProducts();
+
+        assertNotEquals(productsBeforeDel, productsAfterDel);
+        assertTrue(productsBeforeDel.contains(product));
```
no need to check the state before your test method, only after

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/ArrayListProductDaoTest.java:
```diff
@@ -0,0 +1,124 @@
+
+    @Test
+    public void givenProduct_whenDeleteProduct_thenGetProduct() {
+        List<Product> productsBeforeDel = CustomProductDao.getInstance().getProducts();
```
CustomProductDao.getInstance() needs to be done once in setUp()

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/service/CustomProductServiceTest.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Test
+    public void givenProduct_whenDeleteProduct_thenGetProduct() {
+        List<Product> productsBeforeDel = CustomProductService.getInstance().getProducts();
```
you already have productService set in setUp()

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/StringChecker.java:
```diff
@@ -1,14 +1,18 @@
-public class StringChecker {
-    public static boolean containsWords(String string, String search) {
-        if (string == null || search == null) {
+public class StringChecker implements BiPredicate<String, String> {
```
please revert everything what is done for the next task

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/DAO/CustomProductDao.java:
```diff
@@ -14,31 +15,27 @@
 
 public class CustomProductDao implements ProductDao {
-    private static final ReadWriteLock lock = new ReentrantReadWriteLock();//is it a good variant to make locks static?
+    private static final ReadWriteLock lock = new ReentrantReadWriteLock();
```
you can make it non static and for the singleton initialization use the synchronized, in this case we can use synchronized in good for performance way

HidJaba-hub committed:
Fixed some features:
-fixed tests, removed mocking
-namings

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/DAO/CustomProductDao.java:
```diff
@@ -79,11 +61,11 @@ public List<Product> findProducts() {
             if (product == null)
-                throw new ProductException("Product has no data");
+                throw new ProductDefinitionException("Product has no data");
             Optional<Product> oldProduct = getProductById(product.getId());
```
can't we create a deadlock here?
instead of getting getProductById() we can still navigate through stream
P.S. as clarified with documentation, deadlock here can't happen, so no need for changes

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/CustomProductDaoTest.java:
```diff
@@ -0,0 +1,100 @@
+        product = new Product("test", "test", new BigDecimal(100), usd, 10, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg");
+
+        productList = CustomProductDao.getInstance().getProducts();
+        productList.add(product);//as it's a reference we can add products this way to productDao, i suppose, without using save() and other method from test object
```
actually when writing a test you should not rely on real DAO resources
you should mock them
here it's easy to do: create a list of mocked Product-s and via setter add them to dao
it's not a good idea to create tests on real data

HidJaba-hub committed:
Fixed some features:
-fixed tests: add mocking products
-added stream in save

esSoftPalinaTsi commented on src/test/java/com/es/phoneshop/DAO/CustomProductDaoTest.java:
```diff
@@ -2,77 +2,82 @@
+                .forEach(product -> {
+                    when(product.getStock()).thenReturn(10);
+                    when(product.getPrice()).thenReturn(new BigDecimal("10.0"));
+                });//i set only values that i need for checking. am i need to set all the values?
```
no need to set everything

HidJaba-hub committed:
Fixed some features:
-stream removed

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void saveProduct(Product product) {
+        productDao.save(product);
```
what if in dao ProductDefinitionException was thrown?

esSoftPalinaTsi commented on src/main/java/com/es/phoneshop/service/CustomProductService.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Override
+    public void changeState(Product product, boolean state) {
+        CustomProductDao.getInstance().changeChosenState(product, state);
```
productDao.changeChosenState

HidJaba-hub committed:
Fixed some features
